lib/Serializer.ts

```diff
@@ -0,0 +1,120 @@
+import { ObjectType, ModelType } from "./General";
+import { Document } from "./Document";
+
+const defaultSerializer = {
+	"modify": (serialized: ObjectType, original: ObjectType): ObjectType => {
+		return {...original};
+	}
+};
+
+const includeHandler = (document: ModelType<Document>, includeRules: string[], serialized: ObjectType): ObjectType => {
+	return includeRules.reduce((serialized: ObjectType, key: string) => {
+		if (Object.keys(document).includes(key)) {
+			serialized[key] = document[key];
+		}
+
+		return serialized;
+	}, serialized);
+};
+
+const excludeHandler = (document: ModelType<Document>, excludeRules: string[], serialized: ObjectType): ObjectType => {
+	return excludeRules.reduce((serialized: ObjectType, key: string) => {
+		if (Object.keys(serialized).includes(key)) {
+			delete serialized[key];
+		}
+
+		return serialized;
+	}, serialized);
+};
+
+const validateName = (name: string): void => {
+	if (!name || typeof name !== "string") {
+		throw new Error("Field name is required and should be of type string");
+	}
+};
+
+const validateOptions = (options): void => {
+	if (!options || !(Array.isArray(options) || typeof options === "object")) {
+		throw new Error("Field options is required and should be an object or array");
+	}
+};
+
+const cleanAndValidateDocumentsArray = (documentsArray: ModelType<Document>[]): ModelType<Document>[] => {
+	if (!documentsArray || !Array.isArray(documentsArray)) {
+		throw new Error("documentsArray must be an array of document objects");
+	}
+
+	return documentsArray.filter((doc) => typeof doc.serialize === "function");
+};
+
+export class Serializer {
+	#serializers: any;
+	#defaultSerializer: string;
+
+	constructor() {
+		this.#serializers = {_default: defaultSerializer};
+		this.#defaultSerializer = "_default";
+	}
+
+	add(name: string, options): void {
+		validateName(name);
+		validateOptions(options);
+		this.#serializers[name] = options;
+	}
+
+	setDefault(name: string): void {
+		validateName(name);
+		if (Object.keys(this.#serializers).includes(name)) {
+			this.#defaultSerializer = name;
+		}
+	}
+
+	remove(name: string): void {
+		validateName(name);
+		if (Object.keys(this.#serializers).includes(name)) {
+			delete this.#serializers[name];
+		}
+		if (this.#defaultSerializer === name) {
+			this.#defaultSerializer = "_default";
+		}
+	}
+
+	_serializeMany(documentsArray: ModelType<Document>[] = [], nameOrOptions): ObjectType[] {
+		documentsArray = cleanAndValidateDocumentsArray(documentsArray);
+		return documentsArray.map((doc) => doc.serialize(nameOrOptions));
+	}
+
+	_serialize(document: ModelType<Document>, nameOrOptions = this.#defaultSerializer): ObjectType {
+		const inputType = typeof nameOrOptions;
+		let options;
+
+		if (inputType === "string") {
+			options = this.#serializers[nameOrOptions];
+		} else if (Array.isArray(nameOrOptions) || inputType === "object") {
+			options = nameOrOptions;
+		}
```

Comment:
In what situation would neither of these be run @ablankenship10?

Could we just do the following?

```js
if (inputType === "string") {
    options = this.#serializers[nameOrOptions];
} else {
    options = nameOrOptions;
}
```

Otherwise, we need a test for the `else` case where neither of these get run. And I'm not sure what that would look like.

Reply:
Basically in the case you use doc.serialize() without providing a serializer name, SerializerOptions type (array or object), and your don't have a defaultSerializer set, options would remain null and thus throw an error at the validateOptions() line because its not configured properly for use. Another option could be to just return the document as  is instead of failing with an error.

Reply:
@ablankenship10 And in what case would you not have a defaultSerializer set? Just trying to figure out how to write a test for this.

Reply:
If you attempted to use document.serialize() before actually configuring one (actually the included default serializer (at the top of the file) SHOULD always at least be used. This is mainly for handling a case of user error though it may not be possible to hit. Originally the tests I wrote did have complete coverage though so I think I managed to hit it somewhere.

Reply:
@ablankenship10 Got it. Will look into this more after I work on the tests. If it still can't be hit I'll refactor to use else.

I think part of this is also that I changed the default serializer to be a private property. So I don't think it's even accessible for the user to modify anymore.

Reply:
Yea that one is just there so if you don't do anything at all but use serialize() it at least returns the full document. Once you add at least 1 serializer, that one should be used as default unless its removed. Ideally if you only have 1, or a main one, you never have to actually call it by name when calling serialize(). The only purpose of the named ones is in the case where you want something different than the default (i.e. returning a user their own User object with their own account settings configured, vs another users object where that information is not needed/private)